### PR TITLE
URDF Default namespace support

### PR DESCRIPTION
**Pull request by gonzalocasas**
This is a first attempt at supporting URDF docs with a default namespace. I'm not happy with having to pass the default namespace all the way down, but I didn't find a way to traverse the tree upwards from the element instances.

### What type of change is this?

- [x] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tests/compas/robots/test_model.py [hunk 1]
```diff
@@ -116,14 +116,31 @@ def test_robot_urdf_namespaces_to_string():
     # and replace all references to the original with the new.
 
 
-@pytest.mark.skip
-def test_robot_default_namespace(reason="Default parser namespace issues"):
+def test_robot_default_namespace():
```
Comment (beverlylytle) [inline tests/compas/robots/test_model.py:120]: 👍

**Review by beverlylytle — COMMENTED** [review]
I'm concerned about urdfs where the default namespace changes somewhere away from the root, eg
```
<?xml version="1.0" encoding="UTF-8"?>
<robot xmlns="https://kansas.gov" name="SeminamelessLinkRobot">
  <link name="Toto"/>
  <link xmlns="https://somewhere.over.the.rainbow"/>
</robot>

**Comment by gonzalocasas** [comment]
I guess that will not work. But it's not easy to add support for that (or at least I didn't find a clear way to do it). I tried traversing the tree upwards to try that approach, but I couldn't get it to work.

File: src/compas/files/_xml/xml_pre_38.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        while True:
+            data = source.read(65536)
+            if not data:
+                break
+            parser.feed(data)
+
+        return _process_all_events(parser)
+    finally:
+        if close_source:
+            source.close()
```
Comment (beverlylytle) [inline src/compas/files/_xml/xml_pre_38.py:38]: I think we could get some mileage out of abstracting this open-if-necessary-and-read-resulting-file-like-object into a context manager of our own stashed away in `compas.utilities` and then this code could be smushed to `with open_file(source, 'rb', close_source=True) as data: ...`.

File: src/compas/files/_xml/xml_pre_38.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+def xml_from_string(text, tree_parser=None):
+    if tree_parser:
+        raise NotSupportedErr('XML parsing on CPython 3.7 and older does not support a custom tree parser')
+
+    tree_parser = ET.XMLPullParser
+    parser = tree_parser(events=('start', 'end', 'start-ns', 'end-ns'))
```
Comment (beverlylytle) [inline src/compas/files/_xml/xml_pre_38.py:46]: In `_process_all_events`, only `'start'` and `'start-ns'` are used.  So is it necessary to include `'end'` and `'end-ns`'?

File: src/compas/files/_xml/xml_shared.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import io
+import xml.etree.ElementTree as ET
+
+
+def shared_xml_from_file(source, tree_parser, http_response_type):
```
Comment (beverlylytle) [inline src/compas/files/_xml/xml_shared.py:9]: `http_response_type` is a much better name.

File: tests/compas/files/test_xml.py [hunk 5]
```diff
@@ -59,3 +64,73 @@ def test_default_namespace_to_string():
...
+    xml = XML.from_string(
+        """<?xml version="1.0"?><main name="test-xml"></main>"""
+    )
+
+    assert not xml.root.attrib.get('xmlns')
+    assert xml.root.attrib['name'] == 'test-xml'
+
+
+def test_namespace_expansion():
+    xml = XML.from_string(
```
Comment (beverlylytle) [inline tests/compas/files/test_xml.py:127]: would it be worthwhile to have a similar test but `from_file`?

File: tests/compas/robots/test_model.py [hunk 6]
```diff
@@ -116,21 +117,61 @@ def test_robot_urdf_namespaces_to_string():
...
+    assert re.search(b'<link(.*?)(xmlns="https://ita.ethz.ch")(.*?)>', urdf_string)
+
+
+def test_robot_with_prefixed_nested_namespaces_to_string():
+    r = RobotModel.from_urdf_string(
+        """<?xml version="1.0" encoding="UTF-8"?><robot xmlns="https://ethz.ch" name="Acrobot"><link xmlns:custom="https://ita.ethz.ch" name="test"><custom:visual/></link></robot>""")
+    urdf_string = URDF.from_robot(r).to_string(prettify=False)
+    assert b'xmlns="https://ethz.ch"' in urdf_string
+    assert b'xmlns:ns0="https://ita.ethz.ch"' in urdf_string
+    assert b'<ns0:visual' in urdf_string
```
Comment (beverlylytle) [inline tests/compas/robots/test_model.py:174]: i'm surprised these pass in ipy.  maybe the second `assert` of `test_robot_default_namespace_to_string` is too verbose.

**Review by beverlylytle — APPROVED** [review]
Code looks good! Adding another test or the suggested context manager is up to you.

File: tests/compas/robots/test_model.py (hunk 6 above)
Comment (gonzalocasas) [reply tests/compas/robots/test_model.py:174]: indeed! fixed at https://github.com/compas-dev/compas/pull/686/commits/a801ac1c9a28f5df77ca59b9d0e11a43fb7d8101

File: tests/compas/files/test_xml.py (hunk 5 above)
Comment (gonzalocasas) [reply tests/compas/files/test_xml.py:127]: added in https://github.com/compas-dev/compas/pull/686/commits/35b96c5a956a3c02882f6242c913a43f78875b68

File: src/compas/files/_xml/xml_pre_38.py (hunk 3 above)
Comment (gonzalocasas) [reply src/compas/files/_xml/xml_pre_38.py:46]: you're right, fixed in https://github.com/compas-dev/compas/pull/686/commits/ad7dc9f1b64f92f9ee1b5b9bc70875993423301a

File: src/compas/files/_xml/xml_pre_38.py (hunk 2 above)
Comment (gonzalocasas) [reply src/compas/files/_xml/xml_pre_38.py:38]: good point, but I will split that change into its own PR

**Comment by gonzalocasas** [comment]
merging this rabbit hole! 🐰🕳

**Pull request merged**